iroh/src/baomap/flat.rs

```diff
@@ -0,0 +1,1210 @@
+            // check if we have the data cached
+            let data = state.data.get(hash).cloned();
+            Some(Entry {
+                hash: blake3::Hash::from(*hash),
+                entry: EntryData {
+                    data: if let Some(data) = data {
+                        Either::Left(data)
+                    } else {
+                        // get the data path
+                        let path = if entry.owned_data {
+                            // use the path for the data in the default location
+                            self.owned_data_path(hash)
+                        } else {
+                            // use the first external path. if we don't have any
+                            // we don't have a valid entry
+                            entry.external_path()?.clone()
+                        };
+                        Either::Right((path, entry.size))
+                    },
+                    outboard: Either::Left(outboard),
+                },
+            })
+        } else if let Some(entry) = state.partial.get(hash) {
+            let data_path = self.0.options.partial_data_path(*hash, &entry.uuid);
+            let outboard_path = self.0.options.partial_outboard_path(*hash, &entry.uuid);
+            tracing::trace!(
+                "got partial: {} {} {}",
+                hash,
+                entry.size,
+                hex::encode(entry.uuid)
+            );
+            Some(Entry {
+                hash: blake3::Hash::from(*hash),
+                entry: EntryData {
+                    data: Either::Right((data_path, entry.size)),
+                    outboard: Either::Right(outboard_path),
+                },
+            })
+        } else {
+            tracing::trace!("got none {}", hash);
+            None
+        }
+    }
+}
+
+impl ReadonlyStore for Store {
+    fn blobs(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static> {
+        let inner = self.0.state.read().unwrap();
+        let items = inner.complete.keys().copied().collect::<Vec<_>>();
+        Box::new(items.into_iter())
+    }
+
+    fn roots(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static> {
+        unimplemented!()
+    }
+
+    fn validate(&self, _tx: mpsc::Sender<ValidateProgress>) -> BoxFuture<'_, anyhow::Result<()>> {
+        unimplemented!()
+    }
+
+    fn partial_blobs(&self) -> Box<dyn Iterator<Item = Hash> + Send + Sync + 'static> {
+        let lock = self.0.state.read().unwrap();
+        let res = lock.partial.keys().cloned().collect::<Vec<_>>();
+        Box::new(res.into_iter())
+    }
+
+    fn export(
+        &self,
+        hash: Hash,
+        target: PathBuf,
+        mode: ExportMode,
+        progress: impl Fn(u64) -> io::Result<()> + Send + Sync + 'static,
+    ) -> BoxFuture<'_, io::Result<()>> {
+        let this = self.clone();
+        tokio::task::spawn_blocking(move || this.export_sync(hash, target, mode, progress))
```

Comment:
shouldn't these `spawn_blocking` calls use the passed through `runtime::Handle`?

Reply:
I don't have a runtime handle in State currently. I guess I would have to add it and explicitly pass in a runtime in the store constructor.